<a href="https://colab.research.google.com/github/aakankshch/classification/blob/main/Iris_KNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd

# Nearest Neighbors

In this example, we will train a nearest neighbour classifier on the Iris dataset.

## 1. Load data

First, we load the dataset from `sklearn` and print its feature names as well as the target names (labels).

In [ ]:
from sklearn.datasets import load_iris
iris = load_iris()
# print(iris.data) #print(iris)

In [ ]:
print(type(iris))
print(type(iris.data))

<class 'sklearn.utils._bunch.Bunch'>
<class 'numpy.ndarray'>


In [ ]:
#Describe about the dataset
iris.DESCR

'.. _iris_dataset:\n\nIris plants dataset\n--------------------\n\n**Data Set Characteristics:**\n\n    :Number of Instances: 150 (50 in each of three classes)\n    :Number of Attributes: 4 numeric, predictive attributes and the class\n    :Attribute Information:\n        - sepal length in cm\n        - sepal width in cm\n        - petal length in cm\n        - petal width in cm\n        - class:\n                - Iris-Setosa\n                - Iris-Versicolour\n                - Iris-Virginica\n                \n    :Summary Statistics:\n\n    ============== ==== ==== ======= ===== ====================\n                    Min  Max   Mean    SD   Class Correlation\n    ============== ==== ==== ======= ===== ====================\n    sepal length:   4.3  7.9   5.84   0.83    0.7826\n    sepal width:    2.0  4.4   3.05   0.43   -0.4194\n    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)\n    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)\n    ============== ==== =

In [ ]:
# Number of Rows and Columns in the array
print(iris.data.shape)

(150, 4)


In [ ]:
print(iris.feature_names)
print(iris.target_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
['setosa' 'versicolor' 'virginica']


This dataset has 4 features and 3 labels.

Now we get the input matrix X and the label vector Y from the dataset:
*   Each row of the matrix X is an example and each column corresponds to a feature.
*   The label vector Y contains the labels of all examples.

Let's print the sizes of the input matrix X and the label vector Y. Also, print the first 10 examples and labels.

In [ ]:
print(iris.target) #prints the target labels indicting with 0,1 and 2

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


In [ ]:
X, Y = iris.data, iris.target

print(X.shape)
print(Y.shape)

#print(X)
#print(Y)


(150, 4)
(150,)


We see that there are 150 examples in the dataset and each example has 4 features.

### Question: What kind of iris has 5cm x 3cm sepal and 5cm x 2cm petal?

## 2. Split data into train/test sets

In [ ]:
#import the module
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier

In [ ]:
X_train,X_test,y_train,y_test=train_test_split(X,Y,test_size=0.3,random_state=33)

NameError: ignored

In [ ]:
print(X_train.shape) #np.shape(X_train)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)
print(type(X_train))

## 3. Create and train K-NN on train set

In [ ]:
knn_model=KNeighborsClassifier(n_neighbors=7,algorithm='ball_tree',)

In [ ]:
knn_model.fit(X_train,y_train)

KNeighborsClassifier(algorithm='ball_tree', n_neighbors=7)

## 4. Question: What kind of iris has 5cm x 3cm sepal and 3cm x 1cm petal?

In [ ]:
predict=np.array([[5,3,3,1]])

In [ ]:
knn_model.predict(predict) #1 is versicolour

array([1])

## 5. Make predictions on test set and Evaluate the predictions

For classification, we can use accuracy as the metric for evaluation.

In [ ]:
y_pred=knn_model.predict(X_test)
print(y_pred)

[1 1 0 1 2 2 0 0 2 2 2 0 2 1 2 1 1 0 1 2 0 0 2 0 2 2 1 1 2 2 1 1 2 2 2 2 2
 1 1 0 1 1 1 0 0]


In [ ]:
knn_model.score(X_test,y_test) #One way to check the accuracy

0.9777777777777777

In [ ]:
#import the module to check the accuracy
from sklearn.metrics import accuracy_score

In [ ]:
acc=accuracy_score(y_test,y_pred)
print(acc*100)

97.77777777777777


In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        11
           1       0.94      1.00      0.97        15
           2       1.00      0.95      0.97        19

    accuracy                           0.98        45
   macro avg       0.98      0.98      0.98        45
weighted avg       0.98      0.98      0.98        45



In [ ]:
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_test,y_pred))

[[11  0  0]
 [ 0 15  0]
 [ 0  1 18]]


In [ ]:
print('y_test:',y_test)
print('y_pred:',y_pred)

y_test: [1 1 0 1 2 2 0 0 2 2 2 0 2 1 2 1 2 0 1 2 0 0 2 0 2 2 1 1 2 2 1 1 2 2 2 2 2
 1 1 0 1 1 1 0 0]
y_pred: [1 1 0 1 2 2 0 0 2 2 2 0 2 1 2 1 1 0 1 2 0 0 2 0 2 2 1 1 2 2 1 1 2 2 2 2 2
 1 1 0 1 1 1 0 0]


In [ ]:
df=pd.DataFrame({'Actual Values' : y_test,'Predicted Values': y_pred})

In [ ]:
display(df.head(5))

Actual Values  Predicted Values
0              1                 1
1              1                 1
2              0                 0
3              1                 1
4              2                 2